src/d2lfetch-dedupe.js

```diff
@@ -1,51 +1,143 @@
+/**
+ * @typedef {(request: Request) => Promise<Response>} MiddlewareFunc
+ * @typedef {Parameters<ConstructorParameters<typeof Promise>[0]>} PromiseCallbackParams
+ * @typedef {{resolve: PromiseCallbackParams[0]; reject: PromiseCallbackParams[1]}} Resolvers
+ * @typedef {{resolvers: Resolvers, removeAbortListener?: Function}} DeduplicatedRequest
+ *
+ * @typedef {object} InflightRequestEntry
+ * @prop {Request} inflightRequest
+ * @prop {AbortController | null} abortController
+ * @prop {Record<string, DeduplicatedRequest>} dedupedRequests
+ *
+ * @typedef {Record<string, InflightRequestEntry>} InflightRequestInfo
+ */
+
 export class D2LFetchDedupe {
 
 	constructor() {
-		this._inflightRequests = this._inflightRequests || [];
+		/** @type {InflightRequestInfo} */
+		this._inflightRequests = {};
+		this._nextReqId = 0;
 	}
 
+	/**
+	 * @param {Request} request
+	 * @param {MiddlewareFunc} next
+	 *
+	 * @returns {Promise<Response>}
+	 */
 	dedupe(request, next) {
 		if (false === request instanceof Request) {
 			return Promise.reject(new TypeError('Invalid request argument supplied; must be a valid window.Request object.'));
 		}
 
 		if (request.method !== 'GET' && request.method !== 'HEAD' && request.method !== 'OPTIONS') {
-			if (!next) {
-				return Promise.resolve(request);
-			}
 			return next(request);
 		}
 
 		const key = this._getKey(request);
-		if (this._inflightRequests[key]) {
-			this._inflightRequests[key].count++;
-			return this._inflightRequests[key].action;
-		}
+		const reqId = this._nextReqId++;
```

Comment:
The odds of this overflowing is probably really low, but maybe `nextReqId` could be a field of `inFlightRequestEntry`? From what I've parsed so far, it needn't be global, and `dedupedRequests` an array?